pkg/scheduler/schedule_one.go

```diff
@@ -747,27 +751,69 @@ func prioritizeNodes(
 
 // selectHost takes a prioritized list of nodes and then picks one
 // in a reservoir sampling manner from the nodes that had the highest score.
-func selectHost(nodeScores []framework.NodePluginScores) (string, error) {
-	if len(nodeScores) == 0 {
-		return "", fmt.Errorf("empty priorityList")
+// It also returns the top {numberOfNodeScoresToIncludeInScheduleResult} Nodes,
+// and the top of the list will be always the selected host.
+func selectHost(nodeScoreList []framework.NodePluginScores, numberOfNodeScoresToIncludeInScheduleResult int) (string, []framework.NodePluginScores, error) {
+	if len(nodeScoreList) == 0 {
+		return "", nil, fmt.Errorf("empty priorityList")
 	}
-	maxScore := nodeScores[0].TotalScore
-	selected := nodeScores[0].Name
+
+	var h nodeScoreHeap = nodeScoreList
+	heap.Init(&h)
 	cntOfMaxScore := 1
-	for _, ns := range nodeScores[1:] {
-		if ns.TotalScore > maxScore {
-			maxScore = ns.TotalScore
-			selected = ns.Name
-			cntOfMaxScore = 1
-		} else if ns.TotalScore == maxScore {
+	// The top of the heap is the NodeScoreResult with the highest score.
+	selected := heap.Pop(&h).(framework.NodePluginScores)
+	sortedNodeScoreList := make([]framework.NodePluginScores, 0, numberOfNodeScoresToIncludeInScheduleResult)
+
+	// This for-loop will continue until all Nodes with the highest scores get checked for a reservoir sampling,
+	// and sortedNodeScoreList gets {numberOfNodeScoresToIncludeInScheduleResult - 1} elements.
+	for ns := heap.Pop(&h).(framework.NodePluginScores); ns.TotalScore == selected.TotalScore ||
+		len(sortedNodeScoreList) < numberOfNodeScoresToIncludeInScheduleResult-1; ns = heap.Pop(&h).(framework.NodePluginScores) {
+		if ns.TotalScore == selected.TotalScore {
 			cntOfMaxScore++
 			if rand.Intn(cntOfMaxScore) == 0 {
 				// Replace the candidate with probability of 1/cntOfMaxScore
-				selected = ns.Name
+				previousSelected := selected
+				selected = ns
+				ns = previousSelected
```

Comment:
This is not easy to reason about, we can simplify it as follows:

Before the for loop, append the "selected" value to sortedNodeScoreList, and after the loop, replace item 0 with the final selected value.